File: reminder-bot/utils/retry.py
"""Retry utility with exponential backoff for Discord API calls."""

import asyncio
import logging
from typing import Callable, TypeVar, Optional
import discord

logger = logging.getLogger(__name__)

T = TypeVar('T')


async def retry_discord_api(
    func: Callable,
    max_retries: int = 3,
    initial_delay: float = 1.0,
    max_delay: float = 60.0,
    backoff_factor: float = 2.0,
    operation_name: str = "Discord API call"
) -> Optional[T]:
    """
    Retry a Discord API call with exponential backoff.
    Handles rate limits (429) and temporary errors.
    
    Args:
        func: Async function to retry
        max_retries: Maximum number of retry attempts
        initial_delay: Initial delay in seconds
        max_delay: Maximum delay in seconds
        backoff_factor: Multiplier for delay after each retry
        operation_name: Name of operation for logging
    
    Returns:
        Result of function call or None if all retries fail
    """
    delay = initial_delay
    
    for attempt in range(max_retries + 1):
        try:
            return await func()
        except discord.HTTPException as e:
            # Handle rate limits
            if e.status == 429:
                retry_after = e.retry_after if hasattr(e, 'retry_after') else delay
                if attempt < max_retries:
                    logger.warning(f"{operation_name} rate limited. Retrying after {retry_after:.1f}s...")
                    await asyncio.sleep(retry_after)
                    delay = min(delay * backoff_factor, max_delay)
                    continue
            
            # Handle other HTTP errors
            if e.status >= 500 and attempt < max_retries:
                logger.warning(f"{operation_name} server error {e.status} (attempt {attempt + 1}/{max_retries + 1}). Retrying in {delay:.1f}s...")
                await asyncio.sleep(delay)
                delay = min(delay * backoff_factor, max_delay)
                continue
            
            # Client errors (4xx) shouldn't be retried
            logger.error(f"{operation_name} failed with HTTP {e.status}: {e}")
            return None
            
        except (discord.ConnectionClosed, asyncio.TimeoutError) as e:
            # Network errors - retry
            if attempt < max_retries:
                logger.warning(f"{operation_name} connection error (attempt {attempt + 1}/{max_retries + 1}): {e}. Retrying in {delay:.1f}s...")
                await asyncio.sleep(delay)
                delay = min(delay * backoff_factor, max_delay)
                continue
            logger.error(f"{operation_name} failed after {max_retries + 1} attempts: {e}")
            return None
            
        except Exception as e:
            # Unexpected errors
            if attempt == max_retries:
                logger.error(f"{operation_name} failed after {max_retries + 1} attempts: {e}")
                return None
            logger.warning(f"{operation_name} error (attempt {attempt + 1}/{max_retries + 1}): {e}. Retrying in {delay:.1f}s...")
            await asyncio.sleep(delay)
            delay = min(delay * backoff_factor, max_delay)
    
    return None

File: utils/retry.js
/**
 * Retry utility for Discord API calls with exponential backoff
 * Handles rate limits and transient errors
 * 
 * Note: This utility checks error properties dynamically to avoid requiring discord.js
 * at the module level, allowing it to work from shared utils/ directory.
 */

/**
 * Retry a Discord API call with exponential backoff
 * @param {Function} fn - Async function to retry
 * @param {Object} options - Retry options
 * @param {number} options.maxRetries - Maximum retry attempts (default: 3)
 * @param {number} options.initialDelay - Initial delay in ms (default: 1000)
 * @param {number} options.maxDelay - Maximum delay in ms (default: 60000)
 * @param {number} options.backoffFactor - Backoff multiplier (default: 2.0)
 * @param {string} options.operationName - Name of operation for logging
 * @returns {Promise} Result of the function call
 */
async function retryDiscordAPI(fn, options = {}) {
  const {
    maxRetries = 3,
    initialDelay = 1000,
    maxDelay = 60000,
    backoffFactor = 2.0,
    operationName = 'Discord API call',
    logger = null,
  } = options;

  let delay = initialDelay;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      // Check if error is a RateLimitError (check constructor name or error code)
      const isRateLimitError = error.constructor?.name === 'RateLimitError' || 
                               error.code === 'RATELIMIT' ||
                               (error.status === 429 && error.retryAfter !== undefined);
      
      // Check if error is an HTTPError (check constructor name or has status property)
      const isHTTPError = error.constructor?.name === 'HTTPError' || 
                         (error.status !== undefined && error.status >= 400);

      // Handle rate limits
      if (isRateLimitError || (isHTTPError && error.status === 429)) {
        const retryAfter = error.retryAfter ? error.retryAfter * 1000 : delay;
        
        if (attempt < maxRetries) {
          if (logger) {
            logger.warn(`${operationName} rate limited. Retrying after ${retryAfter / 1000}s...`);
          }
          await sleep(retryAfter);
          delay = Math.min(delay * backoffFactor, maxDelay);
          continue;
        }
        
        if (logger) {
          logger.error(`${operationName} rate limited after ${maxRetries + 1} attempts`, error);
        }
        throw error;
      }

      // Handle server errors (5xx)
      if (isHTTPError && error.status >= 500 && error.status < 600) {
        if (attempt < maxRetries) {
          if (logger) {
            logger.warn(`${operationName} server error ${error.status} (attempt ${attempt + 1}/${maxRetries + 1}). Retrying in ${delay / 1000}s...`);
          }
          await sleep(delay);
          delay = Math.min(delay * backoffFactor, maxDelay);
          continue;
        }
        
        if (logger) {
          logger.error(`${operationName} server error after ${maxRetries + 1} attempts`, error);
        }
        throw error;
      }

      // Handle connection errors
      if (error.code === 'ECONNRESET' || error.code === 'ETIMEDOUT' || error.message?.includes('WebSocket')) {
        if (attempt < maxRetries) {
          if (logger) {
            logger.warn(`${operationName} connection error (attempt ${attempt + 1}/${maxRetries + 1}): ${error.message}. Retrying in ${delay / 1000}s...`);
          }
          await sleep(delay);
          delay = Math.min(delay * backoffFactor, maxDelay);
          continue;
        }
        
        if (logger) {
          logger.error(`${operationName} connection error after ${maxRetries + 1} attempts`, error);
        }
        throw error;
      }

      // For other errors, don't retry
      if (logger) {
        logger.error(`${operationName} failed`, error);
      }
      throw error;
    }
  }
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

module.exports = { retryDiscordAPI };
